catroid/src/main/java/org/catrobat/catroid/io/DeviceUserDataAccessor.java

```diff
@@ -0,0 +1,203 @@
+ * http://developer.catrobat.org/license_additional_term
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.catrobat.catroid.io;
+
+import android.support.annotation.VisibleForTesting;
+import android.util.Log;
+
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+
+import org.catrobat.catroid.content.Project;
+import org.catrobat.catroid.content.Scene;
+import org.catrobat.catroid.content.Sprite;
+import org.catrobat.catroid.formulaeditor.UserData;
+
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.FileReader;
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.UUID;
+
+public abstract class DeviceUserDataAccessor {
+
+	public static final String TAG = DeviceUserDataAccessor.class.getSimpleName();
+	private File deviceFile;
+
+	public DeviceUserDataAccessor(File projectDirectory) {
+		deviceFile = new File(projectDirectory, getDeviceFileName());
+	}
+
+	protected abstract Object getLock();
+
+	@VisibleForTesting
+	public void setDeviceFile(File deviceFile) {
+		this.deviceFile = deviceFile;
+	}
+
+	public boolean readUserData(UserData userData) {
+
+		HashMap deviceMap = readMapFromJson();
+
+		if (deviceMap == null) {
+			userData.reset();
+			return false;
+		}
+
+		if (!deviceMap.containsKey(userData.getDeviceKey())) {
+			userData.reset();
+			return false;
+		}
+		userData.setValue(deviceMap.get(userData.getDeviceKey()));
+		return true;
+	}
+
+	public void writeUserData(UserData userData) throws IOException {
+		synchronized (getLock()) {
+			if (!deviceFile.exists()) {
+				deviceFile.createNewFile();
+			}
+
+			HashMap deviceVariableMap = readMapFromJson();
+
+			if (deviceVariableMap == null) {
+				deviceVariableMap = new HashMap<>();
+			}
+
+			deviceVariableMap.put(userData.getDeviceKey(), userData.getValue());
+			writeMapToJson(deviceVariableMap);
+		}
+	}
+
+	public void removeDeviceValue(UserData userData) {
+		synchronized (getLock()) {
+			if (!deviceFile.exists()) {
+				return;
+			}
+
+			try {
+				HashMap deviceVariableMap = readMapFromJson();
+				if (deviceVariableMap == null) {
+					return;
+				}
+
+				deviceVariableMap.remove(userData.getDeviceKey());
+				writeMapToJson(deviceVariableMap);
+			} catch (Exception e) {
+				Log.e(TAG, e.getMessage());
+			}
+		}
+	}
+
+	@VisibleForTesting
+	public HashMap<UUID, Object> readMapFromJson() {
```

Comment:
from SonarLint:


The purpose of the Java Collections API is to provide a well defined hierarchy of interfaces in order to hide implementation details.
Implementing classes must be used to instantiate new collections, but the result of an instantiation should ideally be stored in a variable whose type is a Java Collection interface.
This rule raises an issue when an implementation class:
is returned from a public method.
is accepted as an argument to a public method.
is exposed as a public member.
Noncompliant Code Example
  public class Employees {    private HashSet<Employee> employees = new HashSet<Employee>(); // Noncompliant - "employees" should have type "Set" rather than "HashSet"      public HashSet<Employee> getEmployees() { // Noncompliant      return employees;    }  }  
Compliant Solution
  public class Employees {    private Set<Employee> employees = new HashSet<Employee>(); // Compliant      public Set<Employee> getEmployees() { // Compliant      return employees;    }  }